### PR TITLE
[#1798] Improvement(doc): Upgrade doc for index, auto_increment, mysql_properties

### PR DESCRIPTION
**Pull request by Clearvive**
### What changes were proposed in this pull request?
Upgrade doc for index, auto_increment, mysql_properties

### Why are the changes needed?
Fix: #1798 

### Does this PR introduce _any_ user-facing change?
add index doc

### How was this patch tested?
NO

### CODE REVIEW

**Comment by jerryshao** [comment]
@Clearvive You should also add openapi doc for index.

File: docs/manage-metadata-using-gravitino.md [hunk 1]
```diff
@@ -730,14 +730,15 @@ The following is the table property that Gravitino supports:
...
+| Table sort ordering | Equal to `SORTED BY` in Apache Hive, sort ordering is a method to sort the data in specific ways such as by a column or a function, and then store table data. it will highly improve the query performance under certain scenarios.                                                            | [SortOrder](pathname:///docs/0.3.1/api/java/com/datastrato/gravitino/rel/expressions/sorts/SortOrder.html)               |
+| Table index         | Equal to `UNIQUE KEY` in MySQL , unique key enforces uniqueness of values in one or more columns within a table. It ensures that no two rows have identical values in specified columns, thereby facilitating data integrity and enabling efficient data retrieval and manipulation operations. | [Index](pathname:///docs/0.3.1/api/java/com/datastrato/gravitino/rel/indexes/Index.html)                                 |
```
Comment (yuqi1129) [inline docs/manage-metadata-using-gravitino.md:738]: Equal to `KEY/INDEX` in MySQL

File: docs/manage-metadata-using-gravitino.md [hunk 2]
```diff
@@ -730,14 +730,15 @@ The following is the table property that Gravitino supports:
 
 In addition to the basic settings, Gravitino supports the following features:
```
Comment (yuqi1129) [inline docs/manage-metadata-using-gravitino.md:731]: Can you add a level 4 title here about `partitioning, bucketing, sort ordering and index` here?

File: docs/table-partitioning-bucketing-sort-order-indexes.md [hunk 3]
```diff
@@ -1,7 +1,7 @@
 ---
 title: "Table partitioning, bucketing and sort ordering"
-slug: /table-partitioning-bucketing-sort-order
-date: 2023-12-25
+slug: /table-partitioning-bucketing-sort-order-indexes
+date: 2024-02-01
```
Comment (yuqi1129) [inline docs/table-partitioning-bucketing-sort-order-indexes.md:4]: The `date` here is the modificataion day?

File: docs/table-partitioning-bucketing-sort-order-indexes.md [hunk 4]
```diff
@@ -285,3 +285,136 @@ tableCatalog.createTable(
...
+
+## Indexes
+
+To define an indexed table, you should utilize the following three components to construct a valid indexed table.
+
+- IndexType. Represents the type of index, such as primary key or unique key.
+
+| IndexType     | Description                                                                                                                                       | JSON            | Java                      |
+|---------------|---------------------------------------------------------------------------------------------------------------------------------------------------|-----------------|---------------------------|
+| PRIMARY_KEY   | In a relational database is a field or a combination of fields that uniquely identifies each record in a table.                                   | `PRIMARY_KEY`   | `IndexType.PRIMARY_KEY`   |
```
Comment (yuqi1129) [inline docs/table-partitioning-bucketing-sort-order-indexes.md:297]: I don't see any difference between these two sentences about `PRIMARY_KEY` and `UNIQUE_KEY`.

File: docs/table-partitioning-bucketing-sort-order-indexes.md [hunk 5]
```diff
@@ -285,3 +285,136 @@ tableCatalog.createTable(
...
+| UNIQUE_KEY    | In a relational database is a field or a combination of fields that ensures each record in a table has a distinct value or combination of values. | `UNIQUE_KEY`    | `IndexType.UNIQUE_KEY`    |
+
+- Name. It defines the name of the index.
```
Comment (yuqi1129) [inline docs/table-partitioning-bucketing-sort-order-indexes.md:300]: Please add more detailed about it, for example, `The name of `PRIMARY_KEY` must be `PRIMARY` and ...

File: docs/jdbc-mysql-catalog.md [hunk 6]
```diff
@@ -104,6 +104,14 @@ MySQL doesn't support Gravitino `Boolean` `Fixed` `Struct` `List` `Map` `Timesta
 
 - Doesn't support table properties.
 
+### Table indexes
+
+- Support table PRIMARY_KEY and UNIQUE_KEY.
```
Comment (FANNG1) [inline docs/jdbc-mysql-catalog.md:109]: remove table?

File: docs/jdbc-mysql-catalog.md [hunk 7]
```diff
@@ -104,6 +104,14 @@ MySQL doesn't support Gravitino `Boolean` `Fixed` `Struct` `List` `Map` `Timesta
...
+
+:::caution
```
Comment (FANNG1) [inline docs/jdbc-mysql-catalog.md:111]: notice?

File: docs/table-partitioning-bucketing-sort-order-indexes.md (hunk 3 above)
Comment (mchades) [reply docs/table-partitioning-bucketing-sort-order-indexes.md:4]: I cannot find the date display on the web page, @ch3yne do you know how the date works?

File: docs/open-api/tables.yaml
Comment (mchades) [inline docs/open-api/tables.yaml:1]: You have 2 warnings in open api doc lint, please resolve that:
image

File: docs/manage-metadata-using-gravitino.md [hunk 8]
```diff
@@ -716,6 +716,16 @@ The following types that Gravitino supports:
 
 The related java doc is [here](pathname:///docs/0.3.1/api/java/com/datastrato/gravitino/rel/types/Type.html).
```
Comment (FANNG1) [inline docs/manage-metadata-using-gravitino.md:718]: please add the related content to CreateTable, loadTable, AlterTable

File: docs/jdbc-mysql-catalog.md [hunk 9]
```diff
@@ -100,9 +100,25 @@ Please refer to [Manage Metadata Using Gravitino](./manage-metadata-using-gravit
 MySQL doesn't support Gravitino `Boolean` `Fixed` `Struct` `List` `Map` `Timestamp_tz` `IntervalDay` `IntervalYear` `Union` `UUID` type.
 :::
 
+#### Table column auto-increment
+
+- Supports setting auto-increment.
+:::notice
+MySQL setting an auto-increment column requires simultaneously setting a unique index; otherwise, an error will occur.
```
Comment (mchades) [inline docs/jdbc-mysql-catalog.md:107]: Is this notice mentioned in the official MySQL documentation? If so, could you provide the reference link?

File: docs/open-api/tables.yaml [hunk 10]
```diff
@@ -190,6 +192,13 @@ components:
           default: [ ]
           items:
             $ref: "./partitioning.yaml#/components/schemas/PartitioningSpec"
+        indexes:
+          type: object
+          description: Indexes of the table
+          nullable: true
+          default: { }
```
Comment (mchades) [inline docs/open-api/tables.yaml:199]: default should be an array?

File: docs/open-api/tables.yaml [hunk 11]
```diff
@@ -745,6 +759,59 @@ components:
...
+            "name": "dt",
+            "type": "date",
+            "comment": "dt column comment",
+            "nullable": true
+          }
+        ],
+        "indexes": [
+          {
+            "indexType": "primary_key",
+            "name": "PRIMARY",
```
Comment (mchades) [inline docs/open-api/tables.yaml:796]: Will it work if changed to lowercase?

File: docs/open-api/indexes.yaml [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+        name:
+          type: string
+          description: The name of the index
+        fieldNames:
+          type: array
+          description: The cols of the index
+          items:
+            $ref: "#/components/schemas/FieldNames"
+
+    FieldNames:
```
Comment (mchades) [inline docs/open-api/indexes.yaml:35]: `FieldNames` already defined in  tables.yaml:

https://github.com/datastrato/gravitino/blob/513d59156450eb24d75310784e5bf1506d321ae0/docs/open-api/tables.yaml#L293

File: docs/table-partitioning-bucketing-sort-order-indexes.md (hunk 3 above)
Comment (ch3yne) [reply docs/table-partitioning-bucketing-sort-order-indexes.md:4]:
```
last_update:
  date: 2024-02-02
  author: xxx
```
If you want to display the date and author information, you need to add it in this format.
It will display in the bottom right corner of the document.
image

Comment (mchades) [reply docs/table-partitioning-bucketing-sort-order-indexes.md:4]: As @ch3yne  mentioned, he will address this issue, so I suggest making the modification.

Comment (mchades) [reply docs/table-partitioning-bucketing-sort-order-indexes.md]: Suggest to add `last_update`

File: docs/jdbc-mysql-catalog.md [hunk 13]
```diff
@@ -100,9 +100,27 @@ Please refer to [Manage Metadata Using Gravitino](./manage-metadata-using-gravit
...
+:::
+
 ### Table properties
 
-- Doesn't support table properties.
+- Supports setting table engine property. The default value is `InnoDB`. You can set the table engine property to `MyISAM` 、 `MEMORY`、 `CSV`、 `ARCHIVE`、 `BLACKHOLE`、 `FEDERATED`、 `ndbinfo`、 `MRG_MYISAM`、 `PERFORMANCE_SCHEMA`.
+- Supports setting table auto-increment-offset property. Used to specify the starting value of the auto-increment field.
```
Comment (jerryshao) [inline docs/jdbc-mysql-catalog.md:113]: Can you use a table to list all the supported table properties like other documentation.

File: docs/jdbc-mysql-catalog.md [hunk 14]
```diff
@@ -100,9 +100,27 @@ Please refer to [Manage Metadata Using Gravitino](./manage-metadata-using-gravit
...
+- Doesn't support remove table properties. You can only modify values, not delete properties.
+
+### Table indexes
+
+- Support PRIMARY_KEY and UNIQUE_KEY.
```
Comment (jerryshao) [inline docs/jdbc-mysql-catalog.md:118]: "Supports..."

File: docs/manage-metadata-using-gravitino.md [hunk 15]
```diff
@@ -716,6 +716,16 @@ The following types that Gravitino supports:
...
 
+#### Table column auto-increment
+
+The following is the table column auto-increment Gravitino supports:
+
+| Catalog provider    | Auto-increment mapping                                                                        |
+|---------------------|-----------------------------------------------------------------------------------------------|
+| `jdbc-mysql`        | [MySQL auto-increment mapping](./jdbc-mysql-catalog.md#Table-column-auto-increment)           |
+| `jdbc-postgresql`   | [PostgreSQL auto-increment mapping](./jdbc-postgresql-catalog.md#Table-column-auto-increment) |
```
Comment (jerryshao) [inline docs/manage-metadata-using-gravitino.md:726]: Have you tested if this linked is worked or not?

Comment (jerryshao) [reply docs/manage-metadata-using-gravitino.md:726]: I was thinking that if "T" is uppercase or lowcase.

File: docs/jdbc-mysql-catalog.md (hunk 9 above)
Comment (mchades) [reply docs/jdbc-mysql-catalog.md:107]: Then should we put the link to the doc?

File: docs/open-api/indexes.yaml [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+        - fieldNames
+      properties:
+        indexType:
+          type: string
+          enum:
+            - "primary_key"
+            - "unique_key"
+        name:
+          type: string
+          description: The name of the index
```
Comment (mchades) [inline docs/open-api/indexes.yaml:28]: Need to add `nullable: true` item

File: docs/open-api/tables.yaml [hunk 17]
```diff
@@ -39,6 +39,8 @@ paths:
             examples:
               HiveTableCreate:
                 $ref: "#/components/examples/HiveTableCreate"
+              MysqlTableCreate:
```
Comment (mchades) [inline docs/open-api/tables.yaml:42]: Will the PostgreSQL example be added in another PR?

**Comment by mchades** [comment]
Please rebase on the latest code in order to ensure that the OpenAPI lint passes without any warnings.

File: docs/jdbc-mysql-catalog.md [hunk 18]
```diff
@@ -100,9 +100,32 @@ Please refer to [Manage Metadata Using Gravitino](./manage-metadata-using-gravit
 MySQL doesn't support Gravitino `Boolean` `Fixed` `Struct` `List` `Map` `Timestamp_tz` `IntervalDay` `IntervalYear` `Union` `UUID` type.
 :::
 
+#### Table column auto-increment
+
+- Supports setting auto-increment.
+:::notice
```
Comment (jerryshao) [inline docs/jdbc-mysql-catalog.md:106]: Can you please add a blank line above?

File: docs/jdbc-mysql-catalog.md [hunk 19]
```diff
@@ -100,9 +100,32 @@ Please refer to [Manage Metadata Using Gravitino](./manage-metadata-using-gravit
...
+MySQL setting an auto-increment column requires simultaneously setting a unique index; otherwise, an error will occur.
```
Comment (jerryshao) [inline docs/jdbc-mysql-catalog.md:107]: Is it possible to provide an example here?

File: docs/jdbc-mysql-catalog.md [hunk 20]
```diff
@@ -100,9 +100,32 @@ Please refer to [Manage Metadata Using Gravitino](./manage-metadata-using-gravit
...
+
+- Doesn't support remove table properties. You can only modify values, not delete properties.
+
+### Table indexes
+
+- Supports PRIMARY_KEY and UNIQUE_KEY.
+
+:::notice
+The index name of the PRIMARY_KEY must be PRIMARY
+[Create table index](https://dev.mysql.com/doc/refman/8.0/en/create-table.html)
```
Comment (jerryshao) [inline docs/jdbc-mysql-catalog.md:127]: Also here, can you please provide an example of how to use it.

File: docs/manage-metadata-using-gravitino.md [hunk 21]
```diff
@@ -802,16 +812,19 @@ The following is the table property that Gravitino supports:
 | `jdbc-postgresql`   | [PostgreSQL table property](./jdbc-postgresql-catalog.md#table-properties) | [PostgreSQL type mapping](./jdbc-postgresql-catalog.md#table-column-types) |
 
 
+#### Table partition, bucket, sort order and index
```
Comment (jerryshao) [inline docs/manage-metadata-using-gravitino.md:815]: Table partitioning, bucketing, sort ordering and indexing

File: docs/manage-metadata-using-gravitino.md [hunk 22]
```diff
@@ -802,16 +812,19 @@ The following is the table property that Gravitino supports:
...
+| Table sort ordering | Equal to `SORTED BY` in Apache Hive, sort ordering is a method to sort the data in specific ways such as by a column or a function, and then store table data. it will highly improve the query performance under certain scenarios.                                                             | [SortOrder](pathname:///docs/0.3.1/api/java/com/datastrato/gravitino/rel/expressions/sorts/SortOrder.html)               |
+| Table index         | Equal to `KEY/INDEX` in MySQL , unique key enforces uniqueness of values in one or more columns within a table. It ensures that no two rows have identical values in specified columns, thereby facilitating data integrity and enabling efficient data retrieval and manipulation operations.   | [Index](pathname:///docs/0.3.1/api/java/com/datastrato/gravitino/rel/indexes/Index.html)                                 |
```
Comment (jerryshao) [inline docs/manage-metadata-using-gravitino.md:824]: Can you please update the version to 0.4.0 in link.

Comment (jerryshao) [inline docs/manage-metadata-using-gravitino.md:824]: Also can you please change "Table index" to "Table indexing"?

File: docs/manage-metadata-using-gravitino.md [hunk 23]
```diff
@@ -802,16 +812,19 @@ The following is the table property that Gravitino supports:
...
 
 
-For more information, please see the related document on [partitioning, bucketing, and sorting](table-partitioning-bucketing-sort-order.md).
+For more information, please see the related document on [partitioning, bucketing, sorting, and indexes](table-partitioning-bucketing-sort-order-indexes.md).
```
Comment (jerryshao) [inline docs/manage-metadata-using-gravitino.md:827]: also here.

File: docs/table-partitioning-bucketing-sort-order-indexes.md [hunk 24]
```diff
@@ -285,3 +288,136 @@ tableCatalog.createTable(
 
 </TabItem>
 </Tabs>
+
+## Indexes
```
Comment (jerryshao) [inline docs/table-partitioning-bucketing-sort-order-indexes.md:292]: Can you please change to Indexing?

File: docs/table-partitioning-bucketing-sort-order-indexes.md [hunk 25]
```diff
@@ -285,3 +288,136 @@ tableCatalog.createTable(
...
+        .withDataType(Types.VarCharType.of(1000))
+        .withNullable(true)
+        .build(),
+      ColumnDTO.builder()
+        .withComment("Age of the user")
+        .withName("age")
+        .withDataType(Types.ShortType.get())
+        .withNullable(true)
+        .build(),
+      ColumnDTO.builder()
```
Comment (jerryshao) [inline docs/table-partitioning-bucketing-sort-order-indexes.md:404]: Can you please change to not use xxxDTO.

File: docs/manage-metadata-using-gravitino.md (hunk 8 above)
Comment (FANNG1) [reply docs/manage-metadata-using-gravitino.md:792]: seems not resovled

File: docs/jdbc-mysql-catalog.md [hunk 26]
```diff
@@ -100,9 +103,168 @@ Please refer to [Manage Metadata Using Gravitino](./manage-metadata-using-gravit
...
+```
+
+</TabItem>
+<TabItem value="java" label="Java">
+
+```java
+tableCatalog.createTable(
+        NameIdentifier.of("metalake", "hive_catalog", "schema", "my_mysql_table"),
+        new ColumnDTO[] {
+        ColumnDTO.builder()
```
Comment (jerryshao) [inline docs/jdbc-mysql-catalog.md:156]: Don't use `ColumnDTO`. Also the indention of code block is not correct.

File: docs/jdbc-mysql-catalog.md [hunk 27]
```diff
@@ -100,9 +103,168 @@ Please refer to [Manage Metadata Using Gravitino](./manage-metadata-using-gravit
...
+      "type": "string",
+      "comment": "name column comment",
+      "nullable": true
+    }
+  ],
+  "indexes": [
+    {
+      "indexType": "primary_key",
+      "name": "PRIMARY",
+      "fieldNames": [ [ "id" ] ]
```
Comment (jerryshao) [inline docs/jdbc-mysql-catalog.md:140]: `[["id"]]`

File: docs/jdbc-mysql-catalog.md [hunk 28]
```diff
@@ -100,9 +103,168 @@ Please refer to [Manage Metadata Using Gravitino](./manage-metadata-using-gravit
...
+        Distributions.NONE,
+        new SortOrder[0],
+        new Index[] {
+        Indexes.of(IndexType.PRIMARY_KEY, "PRIMARY", new String[][]{{"id"}}),
+        Indexes.of(IndexType.UNIQUE_KEY, "id_name_uk", new String[][]{{"id"} , {"name"}}),
+        });
+```
+
+</TabItem>
+</Tabs>
```
Comment (jerryshao) [inline docs/jdbc-mysql-catalog.md:267]: Also the indention is not correct.

Besides, you only need to show how to use index, that should be enough. Not necessary to show other codes.

**Review by jerryshao — APPROVED** [review]
